pkg/volume/util/util.go

```diff
@@ -598,3 +598,39 @@ func WriteVolumeCache(deviceMountPath string, exec mount.Exec) error {
 	// For linux runtime, it skips because unmount will automatically flush disk data
 	return nil
 }
+
+// isMultiAttachForbidden checks if attaching this volume to multiple nodes is definitely not allowed/possible.
+// In its current form, this function can only reliably say for which volumes it's definitely forbidden. If it returns
+// false, it is not guaranteed that multi-attach is actually supported by the volume type and we must rely on the
+// attacher to fail fast in such cases.
+// Please see https://github.com/kubernetes/kubernetes/issues/40669 and https://github.com/kubernetes/kubernetes/pull/40148#discussion_r98055047
+func IsMultiAttachForbidden(volumeSpec *volume.Spec) bool {
+	if volumeSpec.Volume != nil {
+		// Check for volume types which are known to fail slow or cause trouble when trying to multi-attach
+		if volumeSpec.Volume.AzureDisk != nil ||
+			volumeSpec.Volume.Cinder != nil {
+			return true
+		}
+	}
+
+	// Only if this volume is a persistent volume, we have reliable information on whether it's allowed or not to
+	// multi-attach. We trust in the individual volume implementations to not allow unsupported access modes
+	if volumeSpec.PersistentVolume != nil {
+		// Check for persistent volume types which do not fail when trying to multi-attach
+		if len(volumeSpec.PersistentVolume.Spec.AccessModes) == 0 {
+			// No access mode specified so we don't know for sure. Let the attacher fail if needed
+			return false
+		}
+
+		// check if this volume is allowed to be attached to multiple PODs/nodes, if yes, return false
+		for _, accessMode := range volumeSpec.PersistentVolume.Spec.AccessModes {
```

Comment:
I pointed it out as a separate issue here: https://github.com/kubernetes/kubernetes/issues/73972#issuecomment-553187337
But the fix was closed as per this comment: https://github.com/kubernetes/kubernetes/pull/85166#discussion_r346651083

I think there's a conflict between the two uses cases of this function. In the original use case of preventing multiple attaches across nodes, we can't forbid multi-attach because the PV spec allows the volume to be used either way. But in this use case, we want to be conservative and not parallelize attaches when the volume is used in ReadWriteOnce mode, which could break the volume.

Is there a better way to tell exactly how the volume is meant to be used at attach time? A combination of pod spec and volume spec should give us this information, so perhaps the volume plugin can tell us the access mode once pod spec and volume spec are known.

The simplest solution is to treat this function as a best-effort check, and delegate the real parallel attach protection to volume plugins. If they are already expected to protect against multi attaches, maybe this is OK.

@jingxu97 @tedyu

Reply:
since AccessMode is optional, I wonder how user actually choose to put this information there. Also if volume plugin supports multiple options, how much information they normally put there.
In current logic, if AccessModes is not set at all, controller will allow attach to multiple nodes and let plugin fail if needed.  I think it is ok to allow parallelization in this case.

Reply:
ACK - leaving the logic as is and letting the plugin handle the protection.